script/analysis/get_dynamical_time_time_parallel.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/evn python
+
+"""get_dynamical_time_time_parallel.py
+Author: Jonah Miller (jonah.maxwell.miller@gmail.com)
+
+This little script computes the integrated characteristic dynamical
+time of a tracer. This is performed time-paralle, using a set of
+pruned tracer files.
+"""
+
+from __future__ import print_function, division
+import sys; sys.dont_write_bytecode = True
+import hdf5_to_dict as io
+import numpy as np
+import os, glob, h5py
+from multiprocessing import Pool
+from units import cgs
+from plot_tracers import get_intersection
+
+SAVE_TAU_IN_TRACERS=False
+SMALL = 1e-20
+
+def _get_id_worker(filename):
+    with h5py.File(filename,'r') as f:
+        dump_id = f['dumptrace_id'][0]
+    return dump_id
+
+def get_dump_ids(filenames,nproc=None):
+    filenames = list(sorted(filenames))
+    p = Pool(processes=nproc)
+    ids = np.array(p.map(_get_id_worker, filenames))
+    return ids
+
+def generate_filename(parent_directory,nstep):
+    template = 'tracers_{:08d}.h5part'
+    fnam = template.format(nstep)
+    return os.path.join(parent_directory, fnam)
+
+def get_td_from_fnam(fnam):
+    hdr,units,data = io.load_tracer(fnam)
+    td = io.TracerData.fromtuple(hdr,units,data)
+    return td
+
+def set_tau_dyn(td_now, td_last):
+    # now and last must share same ids, in same order
+    td_now_ = get_intersection(td_now, td_last).sort()
+    td_last_ = get_intersection(td_last, td_now).sort()
+    # numerator
+    rho_avg = 0.5*(td_now_['rho'] + td_last_['rho'])
+    # denominator
+    drho = td_now_['rho'] - td_last_['rho']
+    dt = td_now_['time'] - td_last_['time']
+    drhodt = np.divide(drho, dt)
+    tau_dyn = np.divide(rho_avg, drhodt + SMALL)
+    # add to td_now_
+    td_now_.data['tau_dyn'] = tau_dyn
+    return td_now_
+
+def _tau_dyn_files_worker(inpt):
+    parent_dir = inpt[0]
+    trace_id_max = inpt[1]
+    dump_ids = inpt[2]
+
+    tau_dyns_int = np.zeros(trace_id_max+1)
+    dt_int = np.zeros_like(tau_dyns_int)
+    
+    for dump_id in dump_ids:
+        print("...",dump_id)
+        fnam_now = generate_filename(parent_dir, dump_id)
+        fnam_last = generate_filename(parent_dir, dump_id - 1)
+        td_now = get_td_from_fnam(fnam_now)
+        td_last = get_td_from_fnam(fnam_last)
+        
+        dt = td_now['time'][0] - td_last['time'][0]
```

Comment:
Should this `dt` be proper time for the tracer particle rather than coordinate time?

Reply:
Uh hmm... I'm not sure.

Reply:
Yes probably but it probably doesn't matter too much.